### PR TITLE
Adding interface mac_config

### PR DESCRIPTION
**Pull request by mhwang408**
@Yi-Tseng Add a new field mac_address into config_params of
message SingletonPort to enable configuration of mac_address.

### CODE REVIEW

File: stratum/hal/lib/common/yang_parse_tree_paths.cc [hunk 1]
```diff
@@ -2455,9 +2455,6 @@ void YangParseTreePaths::AddSubtreeInterfaceFromSingleton(
 
   node = tree->AddNode(GetPath("interfaces")(
       "interface", name)("ethernet")("config")("mac-address")());
-  // TODO(tmadejski) Replace the mock value of config MAC address
-  // (0x112233445566) with a value read from the configuration pushed by the
-  // controller once such field is added.
   SetUpInterfacesInterfaceEthernetConfigMacAddress(0x112233445566, node);
```
Comment (stolsma) [inline stratum/hal/lib/common/yang_parse_tree_paths.cc]: Shouldn't the fixed mac address be changed too? Or do I miss something? (If thats the case just ignore this remark... 😎)

Comment (mhwang408) [reply stratum/hal/lib/common/yang_parse_tree_paths.cc]: I copy these lines to another place and forget to delete them.
Thank you.

File: stratum/hal/lib/common/yang_parse_tree_test.cc [hunk 2]
```diff
@@ -1002,6 +1002,50 @@ TEST_F(YangParseTreeTest,
...
 }
 
+// Check if the 'config/mac-address' OnUpdate action works correctly.
+TEST_F(YangParseTreeTest,
+       InterfacesInterfaceEthernetConfigMacAddressOnUpdateSuccess) {
+  auto path = GetPath("interfaces")(
+      "interface", "interface-1")("ethernet")("config")("mac-address")();
+  static constexpr char kMacAddressYangString[] = "11:22:33:44:55:66";
+  static constexpr uint64 kMacAddressUint64 = 0x112233445566ull;
+  static constexpr char kMacAddressSomethingString[] = "00;00;00;00;00;00";
```
Comment (stolsma) [inline stratum/hal/lib/common/yang_parse_tree_test.cc]: You used ; instead of : as separator, probably wrong?

Comment (mhwang408) [reply stratum/hal/lib/common/yang_parse_tree_test.cc]: Yes, you're right.
But in this case, it's on purpose to test if a user inputs an unformed mac address, he/she will get a "wrong value" response.

File: stratum/hal/lib/common/common.proto [hunk 3]
```diff
@@ -279,6 +279,8 @@ message PortConfigParams {
   ModulationType modulation_type = 5;
   // FEC operational mode
   FecMode fec_mode = 6;
+  // The configured mac address for this port.
+  uint64 mac_address =7;
```
Comment (Yi-Tseng) [inline stratum/hal/lib/common/common.proto]: add a space between `=` and `7`

File: stratum/hal/lib/common/utils.cc [hunk 4]
```diff
@@ -333,6 +333,14 @@ std::string MacAddressToYangString(
                          mac_address & 0xFF);
 }
 
+::google::protobuf::uint64 YangStringToMacAddress(
+    const std::string& yang_string) {
+  std::string tmp_str = yang_string;
+  // Remove colons
+  tmp_str.erase(std::remove(tmp_str.begin(), tmp_str.end(), ':'), tmp_str.end());
+  return strtoul(tmp_str.c_str(), NULL, 16);
```
Comment (Yi-Tseng) [inline stratum/hal/lib/common/utils.cc]: I'd suggest using `strtoull` since `unsigned long` can be 32-bit in some cases.
https://en.cppreference.com/w/cpp/language/types

Comment (mhwang408) [reply stratum/hal/lib/common/utils.cc]: Got it, thanks.

File: stratum/hal/lib/common/common.proto (hunk 3 above)
Comment (mhwang408) [reply stratum/hal/lib/common/common.proto]: Done!

File: stratum/hal/lib/common/yang_parse_tree_test.cc (hunk 2 above)
Comment (pudelkoM) [reply stratum/hal/lib/common/yang_parse_tree_test.cc]: How about giving the malformed MAC string its own test (`InterfacesInterfaceEthernetConfigMacAddressOnUpdateFailure`)? There are more error cases (too long, too short, null bytes, ...) to be checked, which would clutter the success test case.

File: stratum/hal/lib/common/yang_parse_tree_paths.cc [hunk 5]
```diff
@@ -1047,7 +1047,7 @@ void SetUpInterfacesInterfaceEthernetConfigMacAddress(uint64 node_id,
     std::string mac_address_string = typed_val->string_val();
     ::google::protobuf::uint64 mac_address =
         YangStringToMacAddress(mac_address_string);
-    if (mac_address == 0) {
+    if (mac_address_string.compare("00:00:00:00:00:00") !=0 && mac_address == 0) {
```
Comment (pudelkoM) [inline stratum/hal/lib/common/yang_parse_tree_paths.cc]: Use `==` for string comparison where the ordering does not matter.

File: stratum/hal/lib/common/yang_parse_tree_paths.cc [hunk 6]
```diff
@@ -1030,9 +1033,67 @@ void SetUpInterfacesInterfaceEthernetConfigMacAddress(uint64 mac_address,
...
+    // Set the value.
+    auto status = SetValue(node_id, port_id, tree,
+                           &SetRequest::Request::Port::mutable_mac_address,
+                           &MacAddress::set_mac_address, mac_address);
+    if (status != ::util::OkStatus()) {
+      return status;
+    }
+
+    // Update the chassis config
+    ChassisConfig* new_config = config->writable();
```
Comment (pudelkoM) [inline stratum/hal/lib/common/yang_parse_tree_paths.cc]: Minor nitpick: Do you see an easy way to hit this code in a test?
Would improve test coverage and make the coverall check happy.

File: stratum/hal/lib/common/yang_parse_tree_paths.cc (hunk 5 above)
Comment (mhwang408) [reply stratum/hal/lib/common/yang_parse_tree_paths.cc]: Cool...Good point.
I've added the test for InterfacesInterfaceEthernetConfigMacAddressOnUpdateFailure as your suggestion, thanks.

File: stratum/hal/lib/common/admin_service_test.cc [hunk 7]
```diff
@@ -166,7 +166,7 @@ TEST_P(AdminServiceTest, CancelReboot) {
   ::gnoi::system::RebootResponse resp;
   ASSERT_OK(admin_service_->Setup(false));
 
-  req.set_delay(5000000);  // 5ms
+  req.set_delay(10000000);  // 5ms
```
Comment (Yi-Tseng) [inline stratum/hal/lib/common/admin_service_test.cc]: This belongs to another patch, please remove this commit.
Also, need to update the comment

Comment (mhwang408) [reply stratum/hal/lib/common/admin_service_test.cc]: I've removed it and refined the test cases.

File: stratum/hal/lib/common/yang_parse_tree_paths.cc (hunk 6 above)
Comment (mhwang408) [reply stratum/hal/lib/common/yang_parse_tree_paths.cc]: The issue: Fix ChassisConfig in YangParseTreeTest reported by Yi Tseng results in this low coverage problem.

Comment (pudelkoM) [reply stratum/hal/lib/common/yang_parse_tree_paths.cc]: Ok then. 80% should work too.
